extensions/amp-skimlinks/0.1/amp-skimlinks.js

```diff
@@ -0,0 +1,243 @@
+  /**
+   * @param {!AmpElement} element
+   */
+  constructor(element) {
+    super(element);
+
+    /** @private {?../../../src/service/xhr-impl.Xhr} */
+    this.xhr_ = null;
+
+    /** @private {?../../../src/service/ampdoc-impl.AmpDoc} */
+    this.ampDoc_ = null;
+
+    /** @private {?../../../src/service/document-info-impl.DocumentInfoDef} */
+    this.docInfo_ = null;
+
+    /** @private {?../../../src/service/viewer-impl.Viewer} */
+    this.viewer_ = null;
+
+    /** @private {?./link-rewriter/link-rewriter-manager.LinkRewriterManager} */
+    this.linkRewriterService_ = null;
+
+    /** @private {!Object} */
+    this.skimOptions_ = {};
+
+    /** @private {?./tracking.Tracking} */
+    this.trackingService_ = null;
+
+    /** @private {?./affiliate-link-resolver.AffiliateLinkResolver} */
+    this.affiliateLinkResolver_ = null;
+
+    /** @private {?./waypoint.Waypoint} */
+    this.waypoint_ = null;
+
+    /** @private {?./link-rewriter/link-rewriter.LinkRewriter} */
+    this.skimlinksLinkRewriter_ = null;
+
+    /** @private {?string} */
+    this.referrer_ = null;
+  }
+
+  /** @override */
+  buildCallback() {
+    this.xhr_ = Services.xhrFor(this.win);
+    this.ampDoc_ = this.getAmpDoc();
+    this.docInfo_ = Services.documentInfoForDoc(this.ampDoc_);
+    this.viewer_ = Services.viewerForDoc(this.ampDoc_);
+
+    /**
+     * Only one instance of LinkRewriterManager
+     * should be shared between all extensions. Since Skimlinks is the only
+     * extension using it we can instanciate it here for now.
+     */
+    this.linkRewriterService_ = new LinkRewriterManager(this.ampDoc_);
+    this.skimOptions_ = getAmpSkimlinksOptions(this.element, this.docInfo_);
+
+    return whenDocumentReady(
+        /** @type {!Document} */(this.ampDoc_.getRootNode())
+    )
+        .then(() => this.viewer_.getReferrerUrl())
+        .then(referrer => {
+          this.referrer_ = referrer;
+          this.startSkimcore_();
+        });
+  }
+
+  /**
+   * @private
+   */
+  startSkimcore_() {
+    this.trackingService_ = this.initTracking_();
+    this.waypoint_ = new Waypoint(
+        /** @type {!../../../src/service/ampdoc-impl.AmpDoc} */
+        (this.ampDoc_),
+        this.trackingService_,
+        this.referrer_
+    );
+    this.affiliateLinkResolver_ = new AffiliateLinkResolver(
+        /** @type {!../../../src/service/xhr-impl.Xhr} */
+        (this.xhr_),
+        this.skimOptions_,
+        this.waypoint_
+    );
+
+    this.skimlinksLinkRewriter_ = this.initSkimlinksLinkRewriter_();
+  }
+
+
+  /**
+   * Fires impression tracking after beacon API request.
+   * @param {!JsonObject} beaconData - Json response from Beacon API.
+   * @private
+   */
+  sendImpressionTracking_(beaconData) {
+    // Update tracking service with extra info.
+    this.trackingService_.setTrackingInfo({guid: beaconData['guid']});
+    const viewer = Services.viewerForDoc(
+        /** @private {!../../../src/service/ampdoc-impl.AmpDoc} */
```

Comment:
@private should be @type for typecasting

Reply:
Good catch